### PR TITLE
[SPARK-24783][SQL] spark.sql.shuffle.partitions=0 should throw exception

### PR DESCRIPTION
**Pull request by srowen**
## What changes were proposed in this pull request?

Throw an exception if spark.sql.shuffle.partitions=0
This takes over https://github.com/apache/spark/pull/23835

## How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @srowen . Could you preserve the original authorship (@WindCanDie) of https://github.com/apache/spark/pull/23835 please ?

**Comment by srowen** [comment]
Sure, but how do I do that?

**Comment by MaxGekk** [comment]
@srowen I did that by creating new branch from master and merging range of commit from another PR like `git cherry-pick ebe6942^..905e279`. Author of the PR should be main author and you are coauthor.

**Comment by srowen** [comment]
Oh OK, nice trick. The issue here is this is a trivial change and we've already spent a lot of review time and pings on it. The change is, practically, one made by other people now. I generally give credit to new contributors or non-committers where possible; this is a gray area. I'm happy to assign the JIRA. I'm not sure I want to spend another PR on this one given the disproportionate time invested already; contributors also need to be responsive.

**Comment by dongjoon-hyun** [comment]
@srowen . If you are busy, could you change the authorship when you merge this PR? That's just what we need here. It's a single `git` command before you push the merged commit to apache repo.

**Comment by srowen** [comment]
Are you suggesting something else like amending the author tag? I'm open to it if that's all there is. Cherry-picking the commits and resolving conflicts -- OK, I think this is overkill given the original author didn't really author this change. I'll do it if you feel strongly, but I'm not sure that reflects the authorship here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Just in case, how much positive could it be? `Int.MaxValue`?

**Comment by dongjoon-hyun** [comment]
It seems that I'm bothering you too much. :) Sorry about that. Yes, please do that. I think this GitHub PR is okay to be left like this. The real commit in Github/Apache repo is the ground truth.

File: sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: Since we expect it should be a positive value, I would check positive scenario, let's say `Int.MaxValue` and negative one `-1`.

**Review by MaxGekk — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: In theory, yeah. I'm not sure if we check every int parameter to verify it can successfully accept an int; that seems like it's tested by tests of the conf framework. Checking for -1, sure.

**Comment by srowen** [comment]
Done, I amended the author tag. Let's see if that works.

**Review by HyukjinKwon — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
Thank you, @srowen . +1, LGTM, too.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**
